criu/seize.c

```diff
@@ -993,12 +1034,16 @@ int collect_pstree(void)
 
 	pr_debug("Detected cgroup V%d freezer\n", cgroup_v2 ? 2 : 1);
 
-	if (opts.freeze_cgroup && freeze_processes())
-		goto err;
-
-	if (!opts.freeze_cgroup && compel_interrupt_task(pid)) {
-		set_cr_errno(ESRCH);
-		goto err;
+	if (opts.freeze_cgroup && !freeze_cgroup_disabled) {
+		if (freeze_processes())
+			goto err;
+	} else {
+		if (opts.freeze_cgroup && unfreeze_cgroup())
```

Comment:
Correct me if I'm wrong: It looks like here we try to unfreeze the freezer cgroup, which was frozen before running criu ("pre-frozen")?

I don't think this is correct as the "container-manager" which did this "pre-freeze" might do it not just for fun but for some deep synchronization reasons (e.g. filesystem is already checkpointed and by freezing processes it tries to guarantee consistency between fs checkpoint and criu checkpoint).

I believe we should just fail if cgroup is "pre-frozen".

Reply:
Good point. Fixed. Thanks.

Reply:
> I don't think this is correct as the "container-manager" which did this "pre-freeze" might do it not just for fun but for some deep synchronization reasons (e.g. filesystem is already checkpointed and by freezing processes it tries to guarantee consistency between fs checkpoint and criu checkpoint).

This is correct. For example, in the case of CRI-O we freeze the container (https://github.com/cri-o/cri-o/pull/6977) to create a consistent snapshot of the container file system. We would need to find an alternative solution to enable GPU checkpointing.